pkgs/build-support/fetchhg/default.nix

```diff
@@ -3,39 +3,37 @@
   stdenvNoCC,
   mercurial,
 }:
-{
-  name ? null,
-  url,
-  rev ? null,
-  sha256 ? null,
-  hash ? null,
-  fetchSubrepos ? false,
-  preferLocalBuild ? true,
-}:
 
-if hash != null && sha256 != null then
-  throw "Only one of sha256 or hash can be set"
-else
-  # TODO: statically check if mercurial as the https support if the url starts with https.
-  stdenvNoCC.mkDerivation {
-    name = "hg-archive" + (lib.optionalString (name != null) "-${name}");
-    builder = ./builder.sh;
-    nativeBuildInputs = [ mercurial ];
+lib.extendMkDerivation {
+  constructDrv = stdenvNoCC.mkDerivation;
+  excludeDrvArgNames = [ "name" ];
+  extendDrvArgs = (
+    finalAttrs:
+    args@{
+      url,
+      name ? null,
+      sha256 ? null,
+      hash ? null,
+      fetchSubrepos ? false,
+      ...
+    }:
+    if hash != null && sha256 != null then
+      throw "Only one of sha256 or hash can be set"
+    else
+      {
+        name = "hg-archive" + (lib.optionalString (name != null) "-${name}");
+        builder = ./builder.sh;
+        nativeBuildInputs = [ mercurial ];
 
-    impureEnvVars = lib.fetchers.proxyImpureEnvVars;
+        impureEnvVars = lib.fetchers.proxyImpureEnvVars;
 
-    subrepoClause = lib.optionalString fetchSubrepos "S";
+        subrepoClause = lib.optionalString fetchSubrepos "S";
 
-    outputHashAlgo = if hash != null then null else "sha256";
-    outputHashMode = "recursive";
-    outputHash =
-      if hash != null then
-        hash
-      else if sha256 != null then
-        sha256
-      else
-        lib.fakeSha256;
+        insecureArgStr = lib.optionalString (!(lib.hasPrefix "https" url)) "--insecure";
 
-    inherit url rev;
-    inherit preferLocalBuild;
-  }
+        outputHashAlgo = if hash != null then null else "sha256";
+        outputHashMode = "recursive";
+        outputHash = args.hash or args.sha256 or lib.fakeSha256;
```

Comment:
Nit: `lib.fakeSha256` should be replaced with `lib.fakeHash`, and the `outputHashAlgo` should also be specified when `hash = ""`.

We need to decide if we would like to override `hash` or `outputHash`, and reference the corresponding hash variable when deciding on `outputHashAlgo`.